File: golang/q800/q829.go
package q800

import (
	"math"
)

func consecutiveNumbersSum(n int) int {
	doubleN := 2 * n
	f := func(l int) bool {
		return (doubleN-l*l+l)%(2*l) == 0
	}
	lMax := int(math.Sqrt(float64(doubleN)))
	count := 0
	for l := 1; l <= lMax; l++ {
		if f(l) {
			count++
		}
	}
	return count
}

File: hs/src/Q800/Q829.hs
module Q800.Q829
(
    consecutiveNumbersSum
)
where

consecutiveNumbersSum :: Int -> Int
consecutiveNumbersSum n =
    let
        doubleN = 2 * n
        lMax = round . sqrt . fromIntegral $ doubleN
        f l = (doubleN-l*l+l) `mod` (2*l) == 0
    in
        length . filter f $ [1..lMax]
